File: serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "x"
        result: List[str] = []
        
        stack = [root]
        
        while stack:
            
            
            curr_root = stack.pop()
            
            if not curr_root:
                result.append("x")
            else:
                result.append(str(curr_root.val))
                
                stack.append(curr_root.right)
                stack.append(curr_root.left)
    
        return ",".join(result)       
                
                
                        
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        
        s = data.split(",")
        return self.helper(s)
    
    def helper(self, s_list: List[str]):
        
        curr_val = s_list.pop(0)
        
        if curr_val == "x":
            return None
        
        else:
            new_root: "TreeNode" = TreeNode(curr_val)
                
            if s_list:
                new_root.left = self.helper(s_list)
            if s_list:
                new_root.right = self.helper(s_list)
                
            return new_root
            
            
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuffer sb = new StringBuffer();
        for(String s : dfs(root, new ArrayList<String>())){
            sb.append(s+",");
        }
        
        return sb.toString();
            
    }
    
    public static List<String> dfs(TreeNode root, List<String> currList){
        
        if(root == null){ 
            currList.add("x");
            return currList;
        }
        currList.add(Integer.toString(root.val));
        
        dfs(root.left, currList);
        dfs(root.right, currList);
        
        return currList;
        
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        
        Queue<String>queue = new LinkedList<>();
        for(String ch : data.split(",")){
            queue.add(ch);
        }
        return dfs(queue);
    }
    
    public static TreeNode dfs(Queue<String> queue){
        String currVal = queue.remove();
        if(currVal.equals("x")) return null; 
        
        TreeNode curr = new TreeNode(Integer.parseInt(currVal));
        if(!queue.isEmpty())curr.left =  dfs(queue);
        
        if(!queue.isEmpty())curr.right = dfs(queue);
        
        return curr;
        
        
        
    }
    
    
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));